phase2/src/bin/prove.rs

```diff
@@ -0,0 +1,125 @@
+extern crate phase2;
+extern crate bellman_ce;
+extern crate num_bigint;
+extern crate num_traits;
+extern crate exitcode;
+extern crate serde;
+
+use std::fs;
+use std::fs::OpenOptions;
+use num_bigint::BigUint;
+use num_traits::Num;
+use serde::{Deserialize, Serialize};
+use phase2::parameters::MPCParameters;
+use phase2::circom_circuit::CircomCircuit;
+use bellman_ce::groth16::{prepare_verifying_key, create_random_proof, verify_proof};
+use bellman_ce::pairing::{
+    Engine,
+    CurveAffine,
+    ff::{
+        PrimeField,
+    },
+    bn256::{
+        Bn256,
+    },
+};
+
+#[derive(Serialize, Deserialize)]
+struct ProofJson {
+    pub protocol: String,
+    pub pi_a: Vec<String>,
+    pub pi_b: Vec<Vec<String>>,
+    pub pi_c: Vec<String>,
+}
+
+fn main() {
+    let args: Vec<String> = std::env::args().collect();
+    if args.len() != 6 {
+        println!("Usage: \n<circuit.json> <witness.json> <params> <proof.json> <public.json>");
+        std::process::exit(exitcode::USAGE);
+    }
+    let circuit_filename = &args[1];
+    let witness_filename = &args[2];
+    let params_filename = &args[3];
+    let proof_filename = &args[4];
+    let public_filename = &args[5];
+
+    let should_filter_points_at_infinity = false;
+    let rng = &mut rand::XorShiftRng::new_unseeded(); // TODO: change this unsafe unseeded random (!)
+
+    let mut c = CircomCircuit::from_json_file(circuit_filename);
+    c.load_witness_json_file(witness_filename);
+    let input = c.inputs.to_vec();
+
+    let reader = OpenOptions::new()
+        .read(true)
+        .open(params_filename)
+        .expect("unable to open.");
+
+    let mut params = MPCParameters::read(reader, should_filter_points_at_infinity, true).expect("unable to read params");
+
+    params.filter_params();
+    let params = params.get_params();
+
+    println!("Proving...");
+    let proof = create_random_proof(c, &*params, rng).unwrap();
+
+    println!("Checking proof");
+    let pvk = prepare_verifying_key(&params.vk);
+    let result = verify_proof(
+        &pvk,
+        &proof,
+        &input[1..]
+    ).unwrap();
+    assert!(result, "Proof is correct");
+
+    let repr_to_big = |r| {
+        BigUint::from_str_radix(&format!("{}", r)[2..], 16).unwrap().to_str_radix(10)
+    };
+    let repr_to_big2 = |r| {
+        BigUint::from_str_radix(&format!("{}", r)[2..], 16).unwrap().to_str_radix(10)
```

Comment:
Should `repr_to_big` and `repr_to_big2` look identical here. Is that intentional?

Reply:
Yes, but probably there is a better way to do this. Lambdas resolve their type arguments when first used, so I can't apply a single one to both `Fr` and `Fq` types.

Also this conversion code is duplicated in `export_keys` and needs some refactoring

Reply:
Those are good points. If we extract that into a proper function, we can use traits instead. Are you happy to do that @poma? If you give me commit access to the branch I'm happy to help as well.

Reply:
https://github.com/kobigurk/phase2-bn254/pull/4#issuecomment-577653412

Reply:
I don't know rust well enough to do that cleanly so I've added you as a collaborator to my repo